### PR TITLE
HSEARCH-3947 Avoid creating huge arrays for collectors when a query does not have an upper bound limit

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3947

### CODE REVIEW

**Comment by fax4ever** [comment]
the commit https://github.com/hibernate/hibernate-search/pull/2361/commits/3c8bd9a619605a5fb621d60902e7141dafc223e9 is also to check that limited-unbounded queries on large indexes but with small matches still work.

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+	private static void initData() {
+		BulkIndexer bulkIndexer = index.bulkIndexer();
+
+		for ( int i = 1; i <= INDEX_SIZE; i++ ) {
+			String value = i + "";
+			bulkIndexer.add( value, document -> document.addValue( index.binding().field, value ) );
+		}
+
+		bulkIndexer.join();
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:89]: This is what we do elsewhere:

```suggestion
		index.bulkIndexer()
				.add( INDEX_SIZE, i -> documentProvider(
						String.valueOf( i ),
						document -> document.addValue( index.binding().field, String.valueOf( i ) ) )
				) )
				.join();
```

So admittedly the end result is the same.

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	@Test
+	public void fetchAll() {
+		SearchQuery<DocumentReference> query = index.createScope().query()
+				.where( f -> f.match().field( "field" ).matching( "739" ) )
+				.toQuery();
+
+		SearchResult<DocumentReference> documentReferences = query.fetchAll();
+
+		assertThat( documentReferences ).hasDocRefHitsAnyOrder( index.typeName(), "739" );
+		Assertions.assertThat( documentReferences.total().hitCount() ).isEqualTo( 1 );
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:54]:
```suggestion
		assertThat( documentReferences )
				.hasDocRefHitsAnyOrder( index.typeName(), "739" )
				.hasTotalHitCount( 1L );
```

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	public void fetchAll_totalHitCountThreshold() {
+		SearchQuery<DocumentReference> query = index.createScope().query()
+				.where( f -> f.match().field( "field" ).matching( "739" ) )
+				.totalHitCountThreshold( 5 )
+				.toQuery();
+
+		SearchResult<DocumentReference> documentReferences = query.fetchAll();
+
+		assertThat( documentReferences ).hasDocRefHitsAnyOrder( index.typeName(), "739" );
+		Assertions.assertThat( documentReferences.total().hitCount() ).isEqualTo( 1 );
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:67]:
```suggestion
		assertThat( documentReferences )
				.hasDocRefHitsAnyOrder( index.typeName(), "739" )
				.hasTotalHitCount( 1L );
```

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
+/*
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:1]: The last commit has the wrong ticket key.

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		BulkIndexer bulkIndexer = index.bulkIndexer();
+
+		for ( int i = 1; i <= INDEX_SIZE; i++ ) {
+			String value = i + "";
+			bulkIndexer.add( value, document -> document.addValue( index.binding().field, value ) );
+		}
+
+		bulkIndexer.join();
+	}
+
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:91]: I think some tests are missing? You didn't test all the possible cases:

* `hitCount <= 100`
* `100 < hitCount <= 10_000
* `10_000 < hitCount`

**Review by yrodiere — CHANGES_REQUESTED** [review]

**Comment by yrodiere** [comment]
The implementation looks good, but I think some tests are missing? See my comments above.

**Comment by fax4ever** [comment]
@yrodiere thanks for the review, changes I think was done...

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java [hunk 6]
```diff
@@ -91,7 +125,7 @@ private static void initData() {
 		final IndexFieldReference<String> field;
 
 		IndexBinding(IndexSchemaElement root) {
-			field = root.field( "field", c -> c.asString() ).toReference();
+			field = root.field( "field", c -> c.asString().sortable( Sortable.YES ) ).toReference();
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:128]: I'm a bit worried now: was that necessary? It shouldn't be...

Comment (fax4ever) [reply integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneNoLimitSearchIT.java:128]: it is not... I'm removing.

**Review by yrodiere — APPROVED** [review]

**Comment by yrodiere** [comment]
Looks good to me, thanks. Feel free to merge once the build passes.

**Comment by fax4ever** [comment]
Thank you!

**Pull request merged**